### DISCUSSION TITLE
Awesome Smart Contract Examples

### DISCUSSION THREAD

**Discussion by leeduckgo** [NonceGeek/Move-Camp#63] [Materials]
> https://www.smartcontract.engineer/

**Comment by AndyK1999** [comment]
Hey I got a question..
I'm pretty new to this stuff and don't know much. But if I want to build a platform on which you can put orders for buying and selleing erc-20 tokens, and you place a buying order which is active for 30 days and includes the token A you want to buy for an amount off token B. Is their a smart contract that can block the amount of token B in my wallet until the order has been fullfilled or cancelled. Same with selling only viceversa.

**Reply by leeduckgo** [reply]
Sorry, could you explain your question more clearly? Do you want to buy token A by token B, and block the token B when the order is active?

**Reply by Hodgewhj** [reply]
hey,I'll try to answer your question. As I know, there is a smart contract  called HTLC that might  solve your problem. Suppose we have a scenes : xiaoming wants to change token B he owned to token A, and he hopes makes a deal in 30 days.
1. xiaoming can generate a random string s,  and then he calculates a hash lock h which by used hash function [ h = hash(s)) ]
2. xiaoming creates a HTLC contract，and uses hash lock h block the token B he owned and sets the lock time to 30 days 。if the contract is not trigger during the time，the block B will reback to his account.
3. if there is another man xiaozhang has interest to this deal，he can also create a HTLC contract and use the same hash lock h lock Block A he owned and set the lock time. The most important  point is that the time must much shorter than xiaoming's contract time.  
4. if xiaoming reads xiaozhang's contract and he thinks the exchange is fair, then xiaoming can send s to xiaozhang's contract to get the block A.
5. now xiaozhang also has the s,  he can send s to xiaoming's contract and get the block B. As we mentioned in step 3, xiao zhang's contract time is much shorter than xiaoming's contract time,  this condition  guarantees that xiaozhang has enough time to get xiaoming's block after he received the string s.

**Reply by AndyK1999** [reply]
Hey Guys. Thanks for the reply. I think i got it now. Based on your description i got one followup question.
As you mentioned, xioa zhang has to put less time in the contract. Is there some way that he hasn´t to put any time in the contract because its automated. I imagine a button on which you can accept an offer without any further doing. Now that the button is pressed the contract which includes amount Token A and amount Token B executes the exchange automated.

**Reply by AndyK1999** [reply]
has someone ever made something like this?

**Comment by leeduckgo** [comment]
Supported by Fenix
